The physical address translation is missing in this version.
Without this feature, other modules such as vspm cannot work.

diff -bur a/drv/mmngr_drv.c b/drv/mmngr_drv.c
--- a/drv/mmngr_drv.c	2015-11-09 17:46:18.433551846 +0700
+++ b/drv/mmngr_drv.c	2015-11-03 14:01:13.000000000 +0700
@@ -67,6 +67,8 @@
 #include <linux/cdev.h>
 #include <linux/platform_device.h>
 #include <linux/miscdevice.h>
+#include <linux/sched.h>
+#include <linux/mm.h>
 #include <linux/version.h>
 #include <linux/dma-attrs.h>
 #include <linux/dma-contiguous.h>
@@ -475,6 +477,25 @@
 	return ret;
 }
 
+/* change virtual address to physical address */
+static void mm_cnv_addr(struct MM_PARAM *out)
+{
+  struct vm_area_struct *vma;
+  struct mm_struct *mm = current->mm;
+  unsigned long start;
+  unsigned long pfn;
+  int lp = 0;
+
+  for (lp=0; lp<2; lp++) {
+    vma = find_vma(mm, out[lp].user_virt_addr);
+    start = out[lp].user_virt_addr & PAGE_MASK;
+    follow_pfn(vma, start, &pfn);
+    out[lp].hard_addr = pfn << PAGE_SHIFT;
+  }
+  
+  return;
+}
+
 static int map_register(void)
 {
 	if (!request_mem_region(MM_MXI_ADDR, MM_MXI_SIZE, DRVNAME))
@@ -779,6 +801,9 @@
 			goto exit;
 		}
 		break;
+	case MM_IOC_VTOP:   /* change virtual address to physical address */
+		mm_cnv_addr((int __user *)arg);
+		break;
 	default:
 		printk(KERN_ERR "MMD CMD EFAULT\n");
 		ret = -EFAULT;
diff -bur a/include/mmngr_private.h b/include/mmngr_private.h
--- a/include/mmngr_private.h	2015-11-09 17:46:18.437551846 +0700
+++ b/include/mmngr_private.h	2015-11-03 14:01:13.000000000 +0700
@@ -95,6 +95,7 @@
 #define MM_IOC_ALLOC_CO	_IOWR(MM_IOC_MAGIC, 4, struct MM_PARAM)
 #define MM_IOC_FREE_CO	_IOWR(MM_IOC_MAGIC, 5, struct MM_PARAM)
 #define MM_IOC_SHARE	_IOWR(MM_IOC_MAGIC, 6, struct MM_PARAM)
+#define MM_IOC_VTOP	_IOWR(MM_IOC_MAGIC, 7, struct MM_PARAM)     /* virt to phys */
 
 #define DEVNAME		"rgnmm"
 #define DRVNAME		DEVNAME

